File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/AdjustableDatesOrRelativeDateOffset.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class AdjustableDatesOrRelativeDateOffset : ISerialized
    {
        public AdjustableDatesOrRelativeDateOffset(XmlNode xmlNode)
        {
            XmlNode adjustableDatesNode = xmlNode.SelectSingleNode("adjustableDates");
            
            if (adjustableDatesNode != null)
            {
                if (adjustableDatesNode.Attributes["href"] != null || adjustableDatesNode.Attributes["id"] != null) 
                {
                    if (adjustableDatesNode.Attributes["id"] != null) 
                    {
                        adjustableDatesIDRef_ = adjustableDatesNode.Attributes["id"].Value;
                        AdjustableDates ob = new AdjustableDates(adjustableDatesNode);
                        IDManager.SetID(adjustableDatesIDRef_, ob);
                    }
                    else if (adjustableDatesNode.Attributes["href"] != null)
                    {
                        adjustableDatesIDRef_ = adjustableDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        adjustableDates_ = new AdjustableDates(adjustableDatesNode);
                    }
                }
                else
                {
                    adjustableDates_ = new AdjustableDates(adjustableDatesNode);
                }
            }
            
        
            XmlNode relativeDateNode = xmlNode.SelectSingleNode("relativeDate");
            
            if (relativeDateNode != null)
            {
                if (relativeDateNode.Attributes["href"] != null || relativeDateNode.Attributes["id"] != null) 
                {
                    if (relativeDateNode.Attributes["id"] != null) 
                    {
                        relativeDateIDRef_ = relativeDateNode.Attributes["id"].Value;
                        RelativeDateOffset ob = new RelativeDateOffset(relativeDateNode);
                        IDManager.SetID(relativeDateIDRef_, ob);
                    }
                    else if (relativeDateNode.Attributes["href"] != null)
                    {
                        relativeDateIDRef_ = relativeDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        relativeDate_ = new RelativeDateOffset(relativeDateNode);
                    }
                }
                else
                {
                    relativeDate_ = new RelativeDateOffset(relativeDateNode);
                }
            }
            
        
        }
        
    
        #region AdjustableDates_
        private AdjustableDates adjustableDates_;
        public AdjustableDates AdjustableDates_
        {
            get
            {
                if (this.adjustableDates_ != null)
                {
                    return this.adjustableDates_; 
                }
                else if (this.adjustableDatesIDRef_ != null)
                {
                    adjustableDates_ = IDManager.getID(adjustableDatesIDRef_) as AdjustableDates;
                    return this.adjustableDates_; 
                }
                else
                {
                      return this.adjustableDates_; 
                }
            }
            set
            {
                if (this.adjustableDates_ != value)
                {
                    this.adjustableDates_ = value;
                }
            }
        }
        #endregion
        
        public string adjustableDatesIDRef_ { get; set; }
        #region RelativeDate_
        private RelativeDateOffset relativeDate_;
        public RelativeDateOffset RelativeDate_
        {
            get
            {
                if (this.relativeDate_ != null)
                {
                    return this.relativeDate_; 
                }
                else if (this.relativeDateIDRef_ != null)
                {
                    relativeDate_ = IDManager.getID(relativeDateIDRef_) as RelativeDateOffset;
                    return this.relativeDate_; 
                }
                else
                {
                      return this.relativeDate_; 
                }
            }
            set
            {
                if (this.relativeDate_ != value)
                {
                    this.relativeDate_ = value;
                }
            }
        }
        #endregion
        
        public string relativeDateIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/AdjustableDatesOrRelativeDateOffset.cpp
// AdjustableDatesOrRelativeDateOffset.cpp 
#include "AdjustableDatesOrRelativeDateOffset.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

AdjustableDatesOrRelativeDateOffset::AdjustableDatesOrRelativeDateOffset(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //adjustableDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* adjustableDatesNode = xmlNode->FirstChildElement("adjustableDates");

   if(adjustableDatesNode){adjustableDatesIsNull_ = false;}
   else{adjustableDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- adjustableDatesNode , address : " << adjustableDatesNode << std::endl;
   #endif
   if(adjustableDatesNode)
   {
      if(adjustableDatesNode->Attribute("href") || adjustableDatesNode->Attribute("id"))
      {
          if(adjustableDatesNode->Attribute("id"))
          {
             adjustableDatesIDRef_ = adjustableDatesNode->Attribute("id");
             adjustableDates_ = boost::shared_ptr<AdjustableDates>(new AdjustableDates(adjustableDatesNode));
             adjustableDates_->setID(adjustableDatesIDRef_);
             IDManager::instance().setID(adjustableDatesIDRef_,adjustableDates_);
          }
          else if(adjustableDatesNode->Attribute("href")) { adjustableDatesIDRef_ = adjustableDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { adjustableDates_ = boost::shared_ptr<AdjustableDates>(new AdjustableDates(adjustableDatesNode));}
   }

   //relativeDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* relativeDateNode = xmlNode->FirstChildElement("relativeDate");

   if(relativeDateNode){relativeDateIsNull_ = false;}
   else{relativeDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- relativeDateNode , address : " << relativeDateNode << std::endl;
   #endif
   if(relativeDateNode)
   {
      if(relativeDateNode->Attribute("href") || relativeDateNode->Attribute("id"))
      {
          if(relativeDateNode->Attribute("id"))
          {
             relativeDateIDRef_ = relativeDateNode->Attribute("id");
             relativeDate_ = boost::shared_ptr<RelativeDateOffset>(new RelativeDateOffset(relativeDateNode));
             relativeDate_->setID(relativeDateIDRef_);
             IDManager::instance().setID(relativeDateIDRef_,relativeDate_);
          }
          else if(relativeDateNode->Attribute("href")) { relativeDateIDRef_ = relativeDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { relativeDate_ = boost::shared_ptr<RelativeDateOffset>(new RelativeDateOffset(relativeDateNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<AdjustableDates> AdjustableDatesOrRelativeDateOffset::getAdjustableDates()
{
   if(!this->adjustableDatesIsNull_){
        if(adjustableDatesIDRef_ != ""){
             return boost::shared_static_cast<AdjustableDates>(IDManager::instance().getID(adjustableDatesIDRef_));
        }else{
             return this->adjustableDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableDates>();
   }
}
boost::shared_ptr<RelativeDateOffset> AdjustableDatesOrRelativeDateOffset::getRelativeDate()
{
   if(!this->relativeDateIsNull_){
        if(relativeDateIDRef_ != ""){
             return boost::shared_static_cast<RelativeDateOffset>(IDManager::instance().getID(relativeDateIDRef_));
        }else{
             return this->relativeDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<RelativeDateOffset>();
   }
}
}

